### ISSUE TITLE
Make generic equality more efficient

### ISSUE THREAD

**Issue by inkytonik** [inkytonik/cooma#14]
Currently it checks to see if the values are ints, then strs, then records, etc. So there is a linear search there.

Perhaps make the compiler translate the generic operation into type-specific primitives?

**Comment by nhweston** [comment]
There are some problems here:

- If we have `x: A` and `y: A` with `A` a type variable, then without full-scale monomorphisation, the compiler must resort to the generic primitive in translating `x == y`.
- Each field or element of a record, variant, or vector needs to be checked for equality. Should the compiler generate code that specialises the primitive for each recursive check? This will be very hard to implement. Such an implementation at the IR-level might be even slower than what we currently have (since there will be a lot more IR code to interpret).

Is it still worthwhile implementing generic equality without addressing these?

Perhaps we could instead implement a way of identifying the entity type (i.e. integer, string, record, variant, or vector) at runtime instead of having to try each individually?

**Comment by inkytonik** [comment]
Yeah, you are probably right that it's not worth it. Thanks for taking a look.

**Comment by inkytonik** [comment]
BTW, we can come back to a runtime-based approach later, but the issue is not a pressing one.
